### PR TITLE
Code coverage JsonDocument, JsonElement

### PR DESCRIPTION
**Pull request by buyaa-n**

### CODE REVIEW

File: src/System.Text.Json/tests/JsonDocumentTests.cs [hunk 1]
```diff
@@ -411,6 +411,71 @@ public static void ParseJson_UnseekableStream_Async_WithBOM(bool compactData, Te
...
+
+        [Fact]
+        public static void ParseJson_Stream_CanSeekThrows_NullBuffer_CodeCoverage()
+        {
+            byte[] data = Encoding.UTF8.GetBytes(SR.Json400KB);
+            Stream stream = new ThrowOnSeekStream(data);
+            JsonDocument doc = null;
+            Assert.Throws<NotImplementedException>(() =>
+            {
+                doc = JsonDocument.Parse(stream);
```
Comment (hughbe) [inline src/System.Text.Json/tests/JsonDocumentTests.cs]: nit: we don't need to assign the result to a variable. By definition if it throws it can't return anything :)

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs [hunk 2]
```diff
@@ -2244,6 +2244,57 @@ public void WriteBase64String(bool formatted, bool skipValidation)
             }
         }
 
+        [Theory]
+        [InlineData(true, true)]
+        [InlineData(true, false)]
+        [InlineData(false, true)]
+        [InlineData(false, false)]
+        public void WriteBase64StringEscapeLengthBigger(bool formatted, bool skipValidation)
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: This test doesn't seem to involve JsonDocument, JsonElement, or JsonProperty, and so probably does not belong in this change.

File: src/System.Text.Json/tests/JsonDocumentTests.cs [hunk 3]
```diff
@@ -3691,4 +3774,24 @@ private static string GetCompactJson(TestCaseType testCaseType, string jsonStrin
...
+        }
+    }
+
+    public class ThrowOnSeekStream : MemoryStream
+    {
+        public ThrowOnSeekStream(byte[] bytes) : base(bytes)
+        {
+        }
+
+        public override bool CanSeek => throw new NotImplementedException();
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonDocumentTests.cs]: A more unique exception would make the test more identifiable.  Something that isn't generally thrown, such as ApplicationException.  (Then the calling/testing code looks less like we didn't implement JsonDocument.Parse).

File: src/System.Text.Json/tests/JsonDocumentTests.cs [hunk 4]
```diff
@@ -411,6 +411,71 @@ public static void ParseJson_UnseekableStream_Async_WithBOM(bool compactData, Te
...
+            {
+                doc = await JsonDocument.ParseAsync(stream);
+            });
+            Assert.Null(doc);
+        }
+
+        [Fact]
+        public static void ParseJson_Stream_CanSeekThrows_NullBuffer_CodeCoverage()
+        {
+            byte[] data = Encoding.UTF8.GetBytes(SR.Json400KB);
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonDocumentTests.cs]: If we're expecting the test to throw, do we need such a large object?

File: src/System.Text.Json/tests/JsonDocumentTests.cs [hunk 5]
```diff
@@ -411,6 +411,71 @@ public static void ParseJson_UnseekableStream_Async_WithBOM(bool compactData, Te
...
+            Stream stream = new ThrowOnSeekStream(data);
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonDocumentTests.cs]: Streams are IDisposable, so should be in a using statement.

File: src/System.Text.Json/tests/JsonDocumentTests.cs (hunk 1 above)
Comment (bartonjs) [reply src/System.Text.Json/tests/JsonDocumentTests.cs]: I agree.

This test would be better as

```C#
[Fact]
public static void ParseJson_Stream_CanSeekThrows_NullBuffer_CodeCoverage()
{
    using (Stream stream = new ThrowOnSeekStream(new byte[] { 1 } ))
    {
        Assert.Throws<NotImplementedException>(() => JsonDocument.Parse(stream));
    }
}
```

Assuming that a payload of `1` (which is legal JSON) is enough to hit the tested condition.  (I didn't sub out the exception type here, since that's a comment from farther away)

File: src/System.Text.Json/tests/JsonDocumentTests.cs (hunk 4 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonDocumentTests.cs]: No need, at first i was disposing the stream while parsing it (for Stream.Read though not for CanSeek), in that case I was needed bigger, will update

File: src/System.Text.Json/tests/JsonDocumentTests.cs (hunk 1 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonDocumentTests.cs]: Yes, i know it will not return anything and @bartonjs version is enough, I wanted to show that the test is just for code coverage and nothign would happen

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: Yeah it is not related to JsonDocument or JsonElement test coverage, it is beginning of JsonWriter coverage and added here as it is still related to code coverage and to avoid future file conflicts. But anyways i can remove it if you prefer that

File: src/System.Text.Json/tests/JsonDocumentTests.cs (hunk 3 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonDocumentTests.cs]: For me NotImplementedException makes more sense than ApplicationException as that is the most common excepotion could happen/thrown for CanSeek property
> (Then the calling/testing code looks less like we didn't implement JsonDocument.Parse)

Sorry didn't understood this point well. Anyways the test intended for code coverage, not about if JsonDocument.Parse is implemented, but guess it is not what you meant 😕

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs (hunk 2 above)
Comment (bartonjs) [reply src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: Your change title is `Code coverage JsonDocument, JsonElement`.  Nothing should be in the change that is not related to the declared purpose. Assuming there're more code coverage improvements on the way for the writer, this should be held back for those changes.

File: src/System.Text.Json/tests/JsonDocumentTests.cs (hunk 3 above)
Comment (bartonjs) [reply src/System.Text.Json/tests/JsonDocumentTests.cs]: The Assert line in your test is `Assert.Throws<NotImplementedException>(() => JsonDocument.Parse(stream))`.  That looks, at a casual glance, like JsonDocument.Parse is what is throwing the exception; but you're adding coverage for a broken stream implementation.

While NotImplementedException is the most likely exception (since auto-generated code tends to leave that sprinkled about), it's not handled special by the calling code.  So I would use some other exception type, just to indicate that it's an exception from the Stream implementation instead of from the JsonDocument implementation.  It's not a major issue, though.

File: src/System.Text.Json/tests/Utf8JsonWriterTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Utf8JsonWriterTests.cs]: ok, will exclude

File: src/System.Text.Json/tests/JsonDocumentTests.cs (hunk 3 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonDocumentTests.cs]: Got it thanks, updated with more meaningful exceptions

File: src/System.Text.Json/tests/JsonDocumentTests.cs [hunk 6]
```diff
@@ -3691,4 +3745,25 @@ private static string GetCompactJson(TestCaseType testCaseType, string jsonStrin
...
+        {
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            throw new EndOfStreamException();
+        }
+    }
+
+    public class ThrowOnArrayPoolRentStep : MemoryStream
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonDocumentTests.cs]: `ThrowOnCanSeekStream` seems better.  ArrayPoolRent is what you're building coverage for right now, but this type might be useful for other tests later.

Comment (buyaa-n) [reply src/System.Text.Json/tests/JsonDocumentTests.cs]: Good point
